### PR TITLE
Added 5 new workload tests.

### PR DESCRIPTION
**Pull request by adingilloRancher**

### CODE REVIEW

File: tests/v3_api/common.py [hunk 1]
```diff
@@ -221,6 +221,16 @@ def validate_workload_with_sidekicks(p_client, workload, type , ns_name, pod_cou
         assert "running" in pod["status"]["containerStatuses"][1]["state"]
 
 
+def validate_workload_image(p_client, workload, expectedImage):
+    image = p_client.list_workload(workload=workload.id)[0].containers[0].image
+    assert image == expectedImage
+
```
Comment (sangeethah) [inline tests/v3_api/common.py]: @adingilloRancher Include the check for making sure that the pods of the workloads have been started with right image using kubectl command.

File: tests/v3_api/common.py [hunk 2]
```diff
@@ -221,6 +221,16 @@ def validate_workload_with_sidekicks(p_client, workload, type , ns_name, pod_cou
...
+
+def validate_workload_paused(p_client, workload, expectedStatus):
+    workloadStatus = p_client.list_workload(uuid=workload.uuid)[0].paused
+    assert workloadStatus == expectedStatus
```
Comment (sangeethah) [inline tests/v3_api/common.py]: @adingilloRancher Make sure that the pods of the workloads are in the right expected state.

File: tests/v3_api/test_workload.py [hunk 3]
```diff
@@ -82,6 +82,86 @@ def test_wl_cronjob():
...
+    ns = namespace["ns"]
+    con = [{"name": "test1",
+           "image": "sangeetha/testclient"}]
+    name = random_test_name("default")
+    workload = p_client.create_workload(name=name,
+                                        containers=con,
+                                        namespaceId=ns.id)
+    workload = wait_for_wl_to_active(p_client, workload)
+    p_client.update(workload, scale=2)
+    validate_workload(p_client, workload, "deployment", ns.name, 2)
```
Comment (sangeethah) [inline tests/v3_api/test_workload.py]: @adingilloRancher After scaling up workloads , make sure that existing pods are not getting recreated. Check that the pods from before upgrade continue to be in "running" state.

File: tests/v3_api/test_workload.py [hunk 4]
```diff
@@ -82,6 +82,86 @@ def test_wl_cronjob():
...
+    name = random_test_name("default")
+    workload = p_client.create_workload(name=name,
+                                        containers=con,
+                                        namespaceId=ns.id,
+                                        scale=2)
+    workload = wait_for_wl_to_active(p_client, workload)
+    p_client.update(workload, scale=1)
+    wait_for_pods_in_workload(p_client, workload, 1)
+    validate_workload(p_client, workload, "deployment", ns.name)
+
```
Comment (sangeethah) [inline tests/v3_api/test_workload.py]: After scaling down workloads , make sure that existing pods are not getting recreated. Check that the pods from before upgrade continue to be in "running" state.

File: tests/v3_api/test_workload.py [hunk 5]
```diff
@@ -82,6 +82,86 @@ def test_wl_cronjob():
...
+    con = [{"name": "test1",
+           "image": "sangeetha/testclient"}]
+    name = random_test_name("default")
+    workload = p_client.create_workload(name=name,
+                                        containers=con,
+                                        namespaceId=ns.id)
+    workload = wait_for_wl_to_active(p_client, workload)
+    p_client.action(workload, "pause")
+    validate_workload_paused(p_client, workload, True)
+
```
Comment (sangeethah) [inline tests/v3_api/test_workload.py]: @adingilloRancher We should validate the pause/resume functionality in the context of /docs/concepts/workloads/controllers/deployment/#pausing-and-resuming-a-deployment

File: tests/v3_api/common.py [hunk 6]
```diff
@@ -232,6 +232,40 @@ def validate_workload_with_sidekicks(p_client, workload, type, ns_name,
...
+
+def validate_pods_are_running_by_id(expectedpods, workload, ns_name):
+    allpods = execute_kubectl_cmd("get pods -l workload.user.cattle.io/"
+                                  "workloadselector=deployment-{0}-{1}"
+                                  .format(ns_name, workload.name))
+    for expectedpod in expectedpods["items"]:
+        expectedpodname = expectedpod["metadata"]["name"]
+        for currentpod in allpods["items"]:
+            if currentpod["metadata"]["name"] == expectedpodname:
+                assert currentpod["status"]["phase"] == "Running"
```
Comment (sangeethah) [inline tests/v3_api/common.py]: @adingilloRancher You need to make sure that every entry in the expectedpod is present in the currenpod list , if not fail. 
Also you can exit the for loop as soon as the match is found.

File: tests/v3_api/common.py [hunk 7]
```diff
@@ -232,6 +232,40 @@ def validate_workload_with_sidekicks(p_client, workload, type, ns_name,
...
+
+def validate_workload_image(client, workload, expectedImage, ns):
+    workload = client.list_workload(uuid=workload.uuid)[0]
+    assert workload.containers[0].image == expectedImage
+    validate_pod_images(expectedImage, workload, ns.name)
+
+
+def validate_revision_count(revision, expectedLength):
+    assert len(revision) == expectedLength
+
```
Comment (sangeethah) [inline tests/v3_api/common.py]: @adingilloRancher We dont need a separate method for this.  We can do the check directly for the length in places where it is needed.

File: tests/v3_api/test_workload.py [hunk 8]
```diff
@@ -84,6 +84,122 @@ def test_wl_cronjob():
...
 
+def test_wl_upgrade():
+    p_client = namespace["p_client"]
+    ns = namespace["ns"]
+    con = [{"name": "test1",
+           "image": TEST_TARGET_IMAGE}]
+    name = random_test_name("default")
+    workload = p_client.create_workload(name=name,
+                                        containers=con,
+                                        namespaceId=ns.id)
```
Comment (sangeethah) [inline tests/v3_api/test_workload.py]: Create workload with scale > 1.

File: tests/v3_api/common.py [hunk 9]
```diff
@@ -651,6 +686,28 @@ def wait_for_ns_to_become_active(client, ns, timeout=DEFAULT_TIMEOUT):
...
+    get_pods = "get pods -l" + label + " -n " + ns_name
+    pods = execute_kubectl_cmd(get_pods)
+
+    for x in range(0, numofpods-1):
+        pod = pods["items"][x]
+        podimage = pod["spec"]["containers"][0]["image"]
+        while podimage != expectedimage:
+            if time.time() - start > timeout:
+                raise AssertionError(
+                    "Timed out waiting for state to get to active")
```
Comment (bmdepesa) [inline tests/v3_api/common.py]: Is the correct assertion error message?

**Comment by bmdepesa** [comment]
LGTM

**Pull request merged**